### PR TITLE
Replace deprecated cf-mysql release with PXC (not the BOSH release)

### LINKED ISSUE #270: Fix compilation of {brain-tests,cf-mysql,groot-btrfs} release images on the sle15 stemcell
**Is your feature request related to a problem? Please describe.**
After the switch to the sle15 stemcell three release images were left behind due to compilation issues:

* brain-tests (bundled ruby 2.3 icompatible with libopenssl)
* cf-mysql (bundled ruby 2.3 icompatible with libopenssl)
* groot-btrfs (automake, pkgconfig)

**Describe the solution you'd like**
The releases should be changed so that they compile again by updating the problematic packages. There could also be changes required for the stemcell itself.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### LINKED ISSUE #312: HA database configuration for CCDB and UAA
**Is your feature request related to a problem? Please describe.**
Though we're pushing admins to use an external database with kubecf in production, for a better out-of-box experience, we need to include an internal database so that early adopters can simply get kubecf operational quickly.

**Describe the solution you'd like**
@troytop can provide more context but perhaps this is simply a case of adding a helm chart to another database. As for which one, I will defer to him.

### PR DESCRIPTION
**Pull request by aduffeck**
## Description
This pr removes the `cf-mysql` release which is no longer maintained and instead deploys the official [percona-xtradb-cluster chart](https://github.com/helm/charts/tree/master/stable/percona-xtradb-cluster) as a subchart using a [SUSE flavored pxc image](https://github.com/SUSE/pxc-docker).

## Motivation and Context

`cf-mysql` no longer compiles on the SUSE stemcell and it's no longer be worked on so we need another solution for an embedded database.

Closes #270. Closes #312.

## How Has This Been Tested?

I've deployed this successfully to my local kind cluster. All components come up properyl, apps can be pushed successfully. CATS are passing partly, the failures could be related to a lack of resources.

Caveats:
* rotation of the database secrets currently does not work
* the scaling of the pxc deployment has to happen in a different place than the scaling of the other cf components, i.e. `percona-xtradb-cluster.replicas` vs. `sizing.*.instances`. I couldn't find a way to make it configurable from the same place.
* switching between embedded and external database currentlyl happens using two boolean flags one of which has to be set to true. This is due to the fact that conditional subcharts can only be enabled by setting a flag to true. We might want to change the condition of the existing `external_database` flag to `!embedded_database` to make this work using only one flag.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code has security implications.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: deploy/helm/kubecf/assets/operations/instance_groups/database.yaml [hunk 1]
```diff
@@ -1,9 +1,31 @@
-{{- if .Values.features.external_database.enabled -}}
-
-# Remove the database instance group.
+# Remove database instance group and variables.
 - type: remove
   path: /instance_groups/name=database
 
+# Remove database links.
```
Comment (mook-as) [inline deploy/helm/kubecf/assets/operations/instance_groups/database.yaml]: Would it make sense to use an [entaglement](https://github.com/cloudfoundry-incubator/cf-operator/blob/master/docs/entanglements.md#native---bosh) here, with an additional service that has the same selector as the one provided by the helm chart? (As far as I know, we can't modify the existing service to have the correct annotations.)

File: deploy/helm/kubecf/templates/db-seeder.yaml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+          mountPath: "/passwords/locket"
+          readOnly: true
+        - name: credhub-database-password
+          mountPath: "/passwords/credhub"
+          readOnly: true
+        command:
+        - "/bin/bash"
+        - "-ce"
+        - |
+          echo "Starting seeding..."
```
Comment (f0rmiga) [inline deploy/helm/kubecf/templates/db-seeder.yaml]: Could you create a `.sh` file with these contents under https://github.com/SUSE/kubecf/tree/master/deploy/helm/kubecf/assets/scripts/jobs, please? This way we take advantage of syntax highlighting and linting.

File: deploy/helm/kubecf/templates/db-seeder.yaml [hunk 3]
```diff
@@ -0,0 +1,109 @@
+{{- if .Values.features.embedded_database.enabled -}}
+
```
Comment (f0rmiga) [inline deploy/helm/kubecf/templates/db-seeder.yaml]: Could you declare the databases here like in https://github.com/SUSE/kubecf/blob/bde658ee927d59f41158f94bb90dcd6161ce75b4/deploy/helm/kubecf/assets/operations/buildpacks/set_suse_buildpacks.yaml#L9-L18?

This way we can do a loop for declaring the volumes and the mounts from a single source.

File: deploy/helm/kubecf/values.yaml [hunk 4]
```diff
@@ -310,3 +302,32 @@ k8s-host-url: ""
 k8s-service-token: ""
 k8s-service-username: ""
 k8s-node-ca: ""
+
+dependencies:
+  - name: percona-xtradb-cluster
+    enabled: true
+    repository: file://percona-xtradb-cluster
+
+# Configuration for the embedded database solution using the pxc subchart
```
Comment (f0rmiga) [inline deploy/helm/kubecf/values.yaml]:
```suggestion
# Configuration for the embedded database solution using the PXC sub-chart.
```

File: deploy/helm/kubecf/requirements.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+dependencies:
+  - name: percona-xtradb-cluster
+    version: ">= 1.0.3"
+    condition: features.embedded_database.enabled
```
[the remark this answers is not in the document]
Comment (viovanov) [reply deploy/helm/kubecf/requirements.yaml]: 👍

File: deploy/helm/kubecf/assets/operations/instance_groups/database.yaml (hunk 1 above)
Comment (viovanov) [reply deploy/helm/kubecf/assets/operations/instance_groups/database.yaml]: This is used as an "external database" from the point of view of the bosh releases.
So links are not used.

File: deploy/helm/kubecf/templates/database_certs.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+{{- if .Values.features.embedded_database.enabled -}}
+
+{{ $ca := genCA "kubecf" 3650 }}
```
[the remark this answers is not in the document]
Comment (aduffeck) [reply deploy/helm/kubecf/templates/database_certs.yaml]: I tried that but ran into the problem that the pxc chart doesn't implement waiting for the required secrets. It expects them to exist before the chart is installed and we can't guarantee that when the operator is handling them.
See [the documentation](https://github.com/helm/charts/tree/master/stable/percona-xtradb-cluster#manage-certificate-secrets-outside-of-helm) for more information.
Unfortunately I couldn't find a way around that yet..

File: deploy/helm/kubecf/assets/operations/instance_groups/database.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (aduffeck) [reply deploy/helm/kubecf/assets/operations/instance_groups/database.yaml]: This has been there before, but it turns out that you're right. It's not necessary.

File: deploy/helm/kubecf/templates/db-seeder.yaml (hunk 3 above)
Comment (aduffeck) [reply deploy/helm/kubecf/templates/db-seeder.yaml]: done

File: deploy/helm/kubecf/templates/db-seeder.yaml (hunk 2 above)
Comment (aduffeck) [reply deploy/helm/kubecf/templates/db-seeder.yaml]: done

File: deploy/helm/kubecf/requirements.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (aduffeck) [reply deploy/helm/kubecf/requirements.yaml]: done

File: deploy/helm/kubecf/requirements.yaml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+dependencies:
+  - name: percona-xtradb-cluster
+    version: ">= 1.0.3"
```
[the remark this answers is not in the document]
Comment (aduffeck) [reply deploy/helm/kubecf/requirements.yaml]: done

File: deploy/helm/kubecf/assets/operations/instance_groups/database.yaml [hunk 8]
```diff
@@ -309,118 +300,85 @@
...
+  value: *pxc-cluster-ca
+
+- type: replace
+  path: /instance_groups/name=diego-api/jobs/name=silk-controller/properties/database/host
+  value: *pxc-cluster-address
+- type: replace
+  path: /instance_groups/name=diego-api/jobs/name=silk-controller/properties/database/ca_cert?
+  value: *pxc-cluster-ca
 
 {{- $root := . -}}
```
[the remark this answers is not in the document]
Comment (aduffeck) [reply deploy/helm/kubecf/assets/operations/instance_groups/database.yaml]: done

File: deploy/helm/kubecf/assets/operations/instance_groups/database.yaml [hunk 9]
```diff
@@ -309,118 +300,85 @@
   value: false
 
 {{- else -}}
+# Override the address and ca with the one from the pxc subchart
+- type: replace
+  path: /instance_groups/name=diego-api/jobs/name=bbs/properties/diego/bbs/sql/db_host?
+  value: &pxc-cluster-address {{ .Release.Name }}-pxc.{{ .Release.Namespace }}.svc.cluster.local
```
[the remark this answers is not in the document]
Comment (aduffeck) [reply deploy/helm/kubecf/assets/operations/instance_groups/database.yaml]: done

File: deploy/helm/kubecf/templates/database_certs.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (f0rmiga) [reply deploy/helm/kubecf/templates/database_certs.yaml]: Hmm, I'll see if I can come up with an idea for solving this. `genCA` is a blocker for merging this PR.

File: deploy/helm/kubecf/values.yaml [hunk 10]
```diff
@@ -310,3 +301,31 @@ k8s-host-url: ""
...
+percona-xtradb-cluster:
+  replicas: 1
+
+  image:
+    repository: "splatform/percona-xtradb-cluster"
+    tag: 0.0.2
+
+  ssl:
+    enabled: true
+    secret: kubecf.var-pxc-certs
```
Comment (f0rmiga) [inline deploy/helm/kubecf/values.yaml]: This is incorrect. The secret name uses a dynamic value. It's coincidental that it works if the deployer specifies `kubecf` as the installation name using Helm. Is it possible to automatically set this value?

File: deploy/helm/kubecf/templates/db-seeder.yaml [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
```
Comment (f0rmiga) [inline deploy/helm/kubecf/templates/db-seeder.yaml]: For consistency, we should name this file `database_seeder.yaml`.

File: deploy/helm/kubecf/values.yaml (hunk 10 above)
Comment (f0rmiga) [reply deploy/helm/kubecf/values.yaml]: This is fixed by https://github.com/SUSE/kubecf/pull/393/commits/4328e181004a5668963f03d3b7c5fdc8742c4441.

File: deploy/helm/kubecf/templates/db-seeder.yaml (hunk 11 above)
Comment (f0rmiga) [reply deploy/helm/kubecf/templates/db-seeder.yaml]: This file went away with https://github.com/SUSE/kubecf/pull/393/commits/4328e181004a5668963f03d3b7c5fdc8742c4441.

**Comment by f0rmiga** [comment]
The issue with the StatefulSet API version also got fixed by https://github.com/SUSE/kubecf/pull/393/commits/4328e181004a5668963f03d3b7c5fdc8742c4441.

**Comment by f0rmiga** [comment]
@aduffeck Please, take a look into setting `sizing.database.instances=3` and running smoke-tests. I didn't have enough time to investigate why it is very flaky.

File: deploy/helm/kubecf/templates/database_certs.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (f0rmiga) [reply deploy/helm/kubecf/templates/database_certs.yaml]: Fixed by https://github.com/SUSE/kubecf/commit/4328e181004a5668963f03d3b7c5fdc8742c4441.

**Comment by bikramnehra** [comment]
The cf-deployment bump to `12.33` would also bump `pxc-release` from `0.21.0` -> `0.22.0`:
https://github.com/cloudfoundry/cf-deployment/compare/v12.18.0...v12.33.0#diff-f39617caec1965e56248ce1bf556db4dR2423
Although its a minor version release, something to be mindful of while testing this PR.

**Comment by f0rmiga** [comment]
> The cf-deployment bump to `12.33` would also bump `pxc-release` from `0.21.0` -> `0.22.0`:
> [cloudfoundry/cf-deployment@v12.18.0...v12.33.0diff-f39617caec1965e56248ce1bf556db4dR2423](https://github.com/cloudfoundry/cf-deployment/compare/v12.18.0...v12.33.0diff-f39617caec1965e56248ce1bf556db4dR2423)
> Although its a minor version release, something to be mindful of while testing this PR.

@bikramnehra This PR has nothing to do with the BOSH pxc-release.

**Pull request merged**